File: backend/src/main/java/be/technobel/parsemaster/security/JWTProvider.java
package be.technobel.parsemaster.security;

import be.technobel.parsemaster.entity.User;
import be.technobel.parsemaster.enumeration.UserRole;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.Authentication;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Component;
import java.nio.charset.StandardCharsets;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.time.Instant;
import java.util.Comparator;
import java.util.TreeSet;
import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.exceptions.JWTVerificationException;

@Component
public class JWTProvider {
  private final byte[] JWT_SECRET; // 64 bytes
  private final long EXPIRES_AT;   // Lifetime of a token in seconds
  private final UserDetailsService userDetailsService;

  /** Revoked token with its expiration time */
  private record RevokedToken(String token, Instant expiration) {}

  /** Set of revoked tokens */
  private final TreeSet<RevokedToken> revokedTokens = new TreeSet<>(
    Comparator.comparing(RevokedToken::expiration)
  );

  public JWTProvider(
    @Value("${spring.profiles.active}") String activeProfile,
    UserDetailsService userDetailsService
  ) throws NoSuchAlgorithmException {
    this.userDetailsService = userDetailsService;
    if(activeProfile.equals("prod")) {
      this.EXPIRES_AT = 60*60;
      SecureRandom.getInstanceStrong().nextBytes(this.JWT_SECRET = new byte[64]);
    } else {
      this.EXPIRES_AT = 60*60*24;
      this.JWT_SECRET = "L7pRJuXxbttmnWlnH1ivjrfHclzmJrey1JuxIPp5KMRfQgR90I29HkzCDwAqsD2d"
        .getBytes(StandardCharsets.US_ASCII);
    }
  }

  /** Generates a new jwt token from user email and role */
  public String generateToken(String username, UserRole role) {
    return "Bearer " + JWT.create()
      .withSubject(username)
      .withClaim("role", role.toString())
      .withExpiresAt(Instant.now().plusMillis(1000* EXPIRES_AT))
      .sign(Algorithm.HMAC512(JWT_SECRET));
  }

  /** Returns the jwt token from http request (without Bearer prefix) or returns null */
  public String extractToken(HttpServletRequest req) {
    final var header = req.getHeader("Authorization");
    if(header == null || !header.startsWith("Bearer ")) return null;
    return header.replaceFirst("Bearer ", "");
  }

  /**
   * Verifies the validity of the jwt token.
   * @param token bearer jwt token
   */
  public boolean validateToken(String token) {
    try {
      final var jwt = JWT.require(Algorithm.HMAC512(JWT_SECRET))
        .acceptExpiresAt(EXPIRES_AT)
        .withClaimPresence("sub")
        .withClaimPresence("role")
        .build()
        .verify(token);
      final var username = jwt.getSubject();
      final var user = (User)userDetailsService.loadUserByUsername(username);
      if(!user.isEnabled()) return false;
      final var role = jwt.getClaim("role").as(UserRole.class);
      return user.getRole().equals(role);
    }
    catch(JWTVerificationException|UsernameNotFoundException ex) {
      return false;
    }
  }

  /**
   * Decodes the jwt token and authenticate the user.
   * @param token bearer jwt token
   */
  public Authentication createAuthentication(String token) {
    final var jwt = JWT.decode(token);
    final var username = jwt.getSubject();
    final var userDetails = userDetailsService.loadUserByUsername(username);
    return new UsernamePasswordAuthenticationToken(
      userDetails.getUsername(), null, userDetails.getAuthorities()
    );
  }

  /**
   * Revokes the jwt token.
   * @param token bearer jwt token
   */
  public void revoke(String token) {
    if(isNotRevoked(token)) {
      revokedTokens.add(new RevokedToken(token, JWT.decode(token).getExpiresAtAsInstant()));
    }
  }

  /**
   * Returns true if the jwt token is not revoked.
   * @param token bearer jwt token
   * @return true if the token is not revoked
   */
  public boolean isNotRevoked(String token) {
    cleanRevoked();
    return revokedTokens.stream().noneMatch(revoked -> revoked.token.equals(token));
  }

  /** Removes expired jwt token from revoke list */
  private void cleanRevoked() {
    final var now = Instant.now();
    revokedTokens.removeIf(revoked -> revoked.expiration().isBefore(now));
  }
}

File: src/main/kotlin/be/technobel/playzone/pl/security/JWTProvider.kt
package be.technobel.playzone.pl.security

import be.technobel.playzone.dal.models.entities.User
import be.technobel.playzone.dal.models.enums.UserRole
import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import com.auth0.jwt.exceptions.JWTVerificationException
import jakarta.servlet.http.HttpServletRequest
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Component
import java.security.SecureRandom
import java.time.Instant
import java.util.*

@Component
class JWTProvider(private val userDetailsService: UserDetailsService) {
	private companion object {
		val JWT_SECRET = generateSecret(64)

		/** Durée de vie en secondes d'un token */
		const val EXPIRES_AT: Long = 60*60*24 // 60*15: 15 minutes

		private fun generateSecret(length: Int): ByteArray {
			val result = ByteArray(length)
			SecureRandom.getInstanceStrong().nextBytes(result)
			return result
		}
	}

	private data class RevokedToken(val token: String, val expiration: Instant)
	private val revokedTokens = TreeSet<RevokedToken> { a, b -> a.expiration.compareTo(b.expiration) }

	fun generateToken(username: String, role: UserRole): String {
		return "Bearer " + JWT.create()
			.withSubject(username)
			.withClaim("role", role.toString())
			.withExpiresAt(Instant.now().plusMillis(1000*EXPIRES_AT))
			.sign(Algorithm.HMAC512(JWT_SECRET))
	}

	fun extractToken(req: HttpServletRequest): String? {
		val header = req.getHeader("Authorization")
		if(header == null || !header.startsWith("Bearer ")) return null
		return header.replaceFirst("Bearer ", "")
	}

	fun validateToken(token: String): Boolean {
		try {
			val jwt = JWT.require(Algorithm.HMAC512(JWT_SECRET))
				.acceptExpiresAt(EXPIRES_AT)
				.withClaimPresence("sub")
				.withClaimPresence("role")
				.build()
				.verify(token)
			val username = jwt.subject
			val user = userDetailsService.loadUserByUsername(username) as User
			if(!user.isEnabled) return false
			val role = jwt.getClaim("role").`as`(UserRole::class.java)
			return user.role == role
		}
		catch(ex: JWTVerificationException) {
			println(ex.message)
			return false
		}
		catch(ex: UsernameNotFoundException) {
			println(ex.message)
			return false
		}
	}

	fun createAuthentication(token: String): Authentication {
		val jwt = JWT.decode(token)
		val username = jwt.subject
		val userDetails = userDetailsService.loadUserByUsername(username)
		return UsernamePasswordAuthenticationToken(userDetails.username, null, userDetails.authorities)
	}

	fun revoke(token: String) {
		cleanRevoked()
		if(!revokedTokens.any { it.token == token }) {
			revokedTokens.add(RevokedToken(token, JWT.decode(token).expiresAtAsInstant))
		}
	}

	fun isRevoked(token: String): Boolean {
		cleanRevoked()
		return revokedTokens.any { it.token == token }
	}

	private fun cleanRevoked() {
		val now = Instant.now()
		revokedTokens.removeIf { it.expiration < now }
	}
}